### DISCUSSION TITLE
apk file  - WearOS element

### DISCUSSION THREAD

**Discussion by PhilTU** [NightscoutFoundation/xDrip#3480] [Q&A] [answered]
I'm very much not a developer, but more of a volunteer support person. I understand that the WearOS component of xDrip+ lacks an active developer and changes to WearOS itself have made installing xDrip WearOS component on watches more challenging.

One approach suggested for installing to the watch is to separate the WearOS apk out of the main xDrip+ apk and install it. https://xdrip.readthedocs.io/en/latest/smartwatch/wearinstall/#method-3-with-a-computer 

this runs into problems with users not able to find the apk in recent release. I find android_wear_micro_apk.apk in xDrip-plus-20231114-3e86af7 in /res/raw
 
![image](https://github.com/NightscoutFoundation/xDrip/assets/76240903/48fc1ff3-b06f-407a-8635-6787ed18742b)

However in xDrip-plus-20240522-52a73fc it is not there but there is a similar sized file sq.apk in /res but there is no /res/raw

![image](https://github.com/NightscoutFoundation/xDrip/assets/76240903/cdf3e37d-d708-4861-aeb5-056975d9fec0)

Has the WearOS apk been renamed and folders changed ?

**Comment by Navid200** [comment]
Thanks for your help.
I just looked at the delta between the latest Nightly and the one before.
It is actually a lot.  But, quite a lot of it is related to some changes required for the new upcoming BlueJay watch.
The only two changes related to Wear I can find have nothing to do with the location or naming of any wear components.

**Reply by PhilTU** [reply]
Thanks. Is sq.apk the new name for the WearOS component ? Seems to look this way as some of the folders within both are the same :- 
![image](https://github.com/NightscoutFoundation/xDrip/assets/76240903/79bad91c-11e3-43c1-8ed0-5b18b5f87e5a)

**Reply by Navid200** [reply]
Do you see this change between May 22 and May 15?

**Reply by PhilTU** [reply]
No, 19th April has same sq.apk file name and location

as has 10th January

**Reply by Navid200** [reply]
I'm sorry, it is getting hard to understand.

Can you identify two consecutive Nightly releases where one has the problem and the other does not?

**Reply by PhilTU** [reply]
Apologies, I have been working my way back..... 5/12/23 has the raw folder and the android_wear_micro_apk.apk filename as per Patrick's instructions, 7/12/23 does not (and is a slightly smaller apk)

**Reply by Navid200** [reply]
Damn!  That was a serious update.  That was update to AndroidX.  
Let me see

**Reply by Navid200** [reply]
I'm sorry I don't think I can help.  
---  

I am not a senior developer.  In fact, I am not really a developer.  I do contribute.  But, I never did computer programming regularly until I started helping with xDrip.

WearOS is written by including a directory in the repository as shown below.  And this is all my understanding, which could be flawed.
![Screenshot 2024-05-24 095124](https://github.com/NightscoutFoundation/xDrip/assets/51497406/4c5aeec5-2610-4faf-a02b-146662e36e2f)


Do you see the green square (app)?  That's where we have been making changes over the past few years to move xDrip forward.
But, what you see in the red square (wear) is where the wearOS content is.
So, there is a lot of redundant content in the repository because of wearOS the way it has been implemented.  And this is only my limited view.  I could be completely wrong about this.
So, let's say I want to change the way an alert works.  I make my changes in the green square, which affects the phone.  But, there may be a copy of the file I am changing inside the red square.  I personally don't touch anything in the red square because I have no idea how to test it.
I can only assume that there are lots of inconsistencies today between the two squares after so many changes have been made.

When I look at the delta between the two releases you have identified, there are too many changes.
Only someone who understands how wearOS works in xDrip can possibly fix this.

**Reply by PhilTU** [reply]
Yes, I see. I can understand the boat anchor effect of effectively having apps for two related but different operating systems coexisting in the same package.

That's been useful as i think I can point people at a version that may not have the issues that arose  when updates were received that took users beyond 5/12/23 and also when I get the watch I can see if sq.apk is the thing to install onto it as a means to address installation / update issues.

Do you think this may be behind the "null null" complication problem ?

**Reply by Navid200** [reply, marked as answer]
It is possible.  I don't know unless someone who experiences it tries those two versions and confirms that the problem only occurs with December 7, but not with 5.

**Reply by zTeveg** [reply]
Are you actually having a problem installing on the WearOS watch or are you searching info for a future install? Looks as if you don't have a watch at the moment?

I have a Samsung Watch 6 (WearOS 4.0, One Ui 5.0) and a Android 14 phone. I can install on WearOS following method #2 without a problem. [https://xdrip.readthedocs.io/en/latest/smartwatch/wearinstall/#method-2-best]
The only problem is when I update the xDrip Phone app, then the automatic update of the watch does not work. The watch just loops the install process until I restart the watch. Then I just install the update manually with Wear installer 2 (according to method #2 again).
Its a little fiddly to pair the watch/phone but really not a biggie.

For me the watch app works flawlessly. I'm just a little worried that a future update will break functionality and sins there is no active developer on the WearOS component it will not be fixed.
I wonder what everybody else (including the devs) are using. Because is works so good with the Watch 6 and I have not been able to find a better alternative that this, so it puzzles me that there isn't more activity on the WearOS app.

**Reply by PhilTU** [reply]
> Are you actually having a problem installing on the WearOS watch or are you searching info for a future install? Looks as if you don't have a watch at the moment?

Correct, I don't. In trying to help others it became apparent the isntructions were flawed because the file name and folder structure changed.

I have a watch coming this week, I was fairly confident Wear Installer 2 would work even though some people seemed to have issues with it. Thanks for the reassurance.

I think the built in update thing would be broken by the December 2023 changes identified, after all if the file name and location changes what hope is there of the old watch apk knowing this.
